### PR TITLE
Refacto Itinerary with ItineraryLocation + ItineraryCollapsible

### PR DESCRIPTION
**Pull request by Tuxosaurus**
I expect a lot of feedback, don't be shy.

Most of the logic found in the previous Itinerary is now in a new component ItineraryLocation.
ItineraryCollapsible can be used in Itinerary to collapse intermediate stopovers.

kap_itineraryCollapsible

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Unreleased
 
+- **[NEW]** `ItineraryLocation`
+- **[NEW]** `ItineraryCollapsible`
```
Comment (mcellier) [inline CHANGELOG.md]: Personnally, I won't export these two and keep them internal to ui-library (as we do for `Item`)

File: src/itinerary/Itinerary.tsx [hunk 2]
```diff
@@ -52,6 +46,59 @@ const computeRootA11yProps = (ariaLabel?: string, ariaLabelledBy?: string): Root
...
+  }
+
+  const place = isArrival ? places[places.length - 1] : places[0]
+
+  return (
+    <ItineraryLocation
+      place={place}
+      isArrival={isArrival}
+      hasTime={!small && withTime}
+      hasSubLabel={!small && Boolean(place.subLabel)}
```
Comment (mcellier) [inline src/itinerary/Itinerary.tsx]: `isEmpty`

File: src/itinerary/Itinerary.tsx [hunk 3]
```diff
@@ -78,132 +131,26 @@ const Itinerary = ({
...
-                tag={TextTagType.PARAGRAPH}
-                display={TextDisplayType.CAPTION}
-                textColor={color.fadedText}
-              >
-                {toAddon}
-              </Text>
-            </div>
-          </li>
+        {renderAddon('from', fromAddon, fromAddonAriaLabel)}
+        {renderDepartureOrArrival(places, false, small, withTime, false)}
```
Comment (mcellier) [inline src/itinerary/Itinerary.tsx]: Not a fan of this naming, it's kind of misleading: we render an addon (from), followed by a departure OR an arrival. IMO we should directly call a `renderLocation` with the right place as a parameter.

File: src/itineraryCollapsible/ItineraryCollapsible.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    }
+  }
+
+  removeListeners() {
+    if (canUseEventListeners) {
+      document.removeEventListener('keydown', this.handleKeydown)
+    }
+  }
+
+  handleKeydown = (event: KeyboardEvent) => {
```
Comment (mcellier) [inline src/itineraryCollapsible/ItineraryCollapsible.tsx]: noice!

File: src/itineraryCollapsible/index.tsx [hunk 5]
```diff
@@ -0,0 +1,83 @@
+import styled from 'styled-components'
+import { color, font, radius, space, componentSizes } from '_utils/branding'
+
+import ItineraryCollapsible from './ItineraryCollapsible'
+
+const minHeight = '32px' // space.l for the minimum content + 2*space.m for vertical padding
+const transitionDelay = '200ms'
```
Comment (mcellier) [inline src/itineraryCollapsible/index.tsx]: Don't we have a default & standard `transitionDelay`?

File: CHANGELOG.md (hunk 1 above)
Comment (mcellier) [reply CHANGELOG.md]: Meaning that i'll:
- put these two components inside `src/itinerary`
- won't create a story about them. 

We should use them though `Itinerary` only, not as standalone components

**Comment by wakooka** [comment]
image
It's not aligned in Chrome

File: src/itineraryCollapsible/index.tsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+const StyledItineraryCollapsible = styled(ItineraryCollapsible)`
+  & {
+    position: relative;
+    display: block;
+    list-style-type: none;
+    cursor: row-resize;
```
[the remark this answers is not in the document]
Comment (Tuxosaurus) [reply src/itineraryCollapsible/index.tsx]: Product also advises to use `cursor: pointer` so I'll change that

File: src/itinerary/Itinerary.tsx [hunk 7]
```diff
@@ -78,132 +131,26 @@ const Itinerary = ({
...
+            ariaLabel={collapsedAriaLabel}
+          />
         )}
+        {!isCollapsible &&
+          intermediatePlaces.length > 0 &&
+          intermediatePlaces.map(place => (
+            <ItineraryLocation
+              place={place}
+              hasTime={!small && withTime}
+              hasSubLabel={!small && Boolean(place.subLabel)}
```
Comment (wakooka) [inline src/itinerary/Itinerary.tsx]: I think you need to add a key attribute here.

File: src/itineraryCollapsible/ItineraryCollapsible.tsx [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  readonly className?: Classcat.Class
+  readonly label?: string
+  readonly ariaLabel?: string
+}
+
+interface ItineraryCollapsibleState {
+  readonly collapsed: boolean
+}
+
+const collapsibleLocationSize = 36 // default size of a collapsible ItineraryLocation
```
Comment (Tuxosaurus) [inline src/itineraryCollapsible/ItineraryCollapsible.tsx]: I'd rather avoid that but I couldn't find a better way to make the transition work from small/hidden size to full size.
See https://css-tricks.com/using-css-transitions-auto-dimensions/ for more explanations

File: src/itineraryCollapsible/ItineraryCollapsible.tsx (hunk 4 above)
Comment (Tuxosaurus) [reply src/itineraryCollapsible/ItineraryCollapsible.tsx]: copy/pasted...

File: src/itineraryCollapsible/ItineraryCollapsible.tsx [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  removeListeners() {
+    if (canUseEventListeners) {
+      document.removeEventListener('keydown', this.handleKeydown)
+    }
+  }
+
+  handleKeydown = (event: KeyboardEvent) => {
+    if (event.keyCode === KEYCODES.ENTER || event.keyCode === KEYCODES.SPACEBAR) {
```
[the remark this answers is not in the document]
Comment (Tuxosaurus) [reply src/itineraryCollapsible/ItineraryCollapsible.tsx]: Yes couldn't use a real button

File: src/itineraryCollapsible/index.tsx [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    height: ${minHeight};
+    background-color: ${color.white};
+  }
+
+  & .kirk-itineraryCollapsible-collapsed::before,
+  & .kirk-itineraryCollapsible-collapsed::after {
+    content: '';
+    position: absolute;
+    top: 0;
+    left: 2px;
```
[the remark this answers is not in the document]
Comment (Tuxosaurus) [reply src/itineraryCollapsible/index.tsx]: Yes, there are still some magic numbers in this file, I'll see what I can do

File: src/itineraryCollapsible/ItineraryCollapsible.tsx [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+          className="kirk-itineraryCollapsible-extended"
+          style={{
+            height: `${places.length * collapsibleLocationSize}px`, // used for the animation
+          }}
+        >
+          {places.map(place => (
+            <ItineraryLocation
+              place={place}
+              isSmall
+              key={`${computeKeyFromPlace(place)}-collapsible`}
```
Comment (Tuxosaurus) [inline src/itineraryCollapsible/ItineraryCollapsible.tsx]: I think there's still a warning issue on keys, this was an attempt to fix it but normally the ItineraryLocation should compute its own key

File: CHANGELOG.md (hunk 1 above)
Comment (mcellier) [reply CHANGELOG.md]: Remove internal components from changelog

File: src/itinerary/Itinerary.tsx (hunk 2 above)
Comment (mcellier) [reply src/itinerary/Itinerary.tsx]: ⬆️

File: stories/index.tsx [hunk 12]
```diff
@@ -26,6 +26,8 @@ import '../src/itemInfo/story'
 import '../src/itemRadio/story'
 import '../src/itemsList/story'
 import '../src/itinerary/story'
+import '../src/itineraryCollapsible/story'
```
Comment (mcellier) [inline stories/index.tsx]: Remove these

File: src/_utils/itineraryCollapsible/story.tsx [hunk 13]
```diff
@@ -0,0 +1,56 @@
+import React from 'react'
```
Comment (mcellier) [inline src/_utils/itineraryCollapsible/story.tsx]: We shouldn't have stories for internal components. They're exposed though `Itinerary`

File: src/_utils/itineraryLocation/index.tsx [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    background-color: ${color.tapHighlight};
+  }
+
+  /* city */
+  & time {
+    min-width: var(--time-width);
+  }
+
+  & .kirk-itineraryLocation-chevron {
+    display: flex;
```
Comment (Tuxosaurus) [inline src/_utils/itineraryLocation/index.tsx]: I would add `align-self: center;` so the icon appears at the middle of the location item

**Pull request merged**